docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+# Introduction
+
+This document proposes a Registration API which allows for components
+of Open MCT to communicate with one another using reusable, internally
+standardized patterns.
+
+# Developer Use Cases
+
+1. Extending and maintaining Open MCT itself.
+2. Adapting and customizing Open MCT for use in specific missions.
+3. Developing features for use with Open MCT across multiple different
+   missions.
+
+# Scope
+
+As demonstrated by the existing APIs, writing plugins is sufficient to
+satisfy the three use cases above in the majority of cases. The only feature
+which is known to be unsatisfiable by plugins is plugin support itself.
+
+As such, prefer to keep "plugin-external" components small and simple, to
+keep the majority of development in plugins (or plugin-like components.)
+
+The "registration API" described in this document is limited to that scope:
+It describes classes and patterns that can allow plugins to interact,
+while making minimal assumptions about what specific functionality is to
+be implemented in these plugins.
+
+By analogy to current API, this set of functionality will effectively
+replace the [Framework Layer](http://nasa.github.io/openmctweb/guide/#framework)
+of Open MCT.
+
+# Problems to Address
+
+1. Dependencies between plugins are implicit; a plugin may fail if its
+   required dependencies are not included, without any clear indication
+   of why this failure has occurred. Significant familiarity with
+   Open MCT is typically required to debug in these circumstances.
+2. Extension points are often implicit; no specific plugin is
+   identifiably responsible for defining any specific extension category
+   or composite service. These are instead paired by string matching.
+   This makes it difficult to follow how the application is initialized
+   and how objects are passed around at run-time, and creates an
+   additional documentation burden, as these named extension points do not
+   fit into more standard API documentation.
+3. Reuse of components between plugins is limited. Exposing base classes
+   from one plugin and reusing them from another is overly-difficult.
+
+# Principles
+
+1. The Registration API is exposed as a set of classes in one or more
+   namespaces. This supports reuse and extension using standard,
+   well-known object-oriented patterns. A composition-oriented style
+   is still supported and encouraged, but not enforced.
+2. Extension points should be expressed as objects with known interfaces.
+   (More specifically, they should be _expressible_ in this fashion;
+   it is out of scope for the Registration API to expose any specific
```

Comment:
@akhenry wrote:

> I actually agree with @larkin. If extension points are dispersed throughout modules it makes it more difficult to discover the extension points available. Of course the developer docs will document them, but it's much simpler if the code itself does it. One of the advantages of an imperative style is that we can declare via an API itself, or through the API docs what extension points are available. Framework should also allow registration of new extension points in addition to this.

> Perhaps could be achieved though through a single code module exposing all core extension points. Could be in the namespace Mct.extensions or something

I'm of two minds about this:

* __Modularity!__ Plugins should have clear dependencies. Sticking all extension points in one place creates a [ginormous context object](http://misko.hevery.com/2008/10/27/pass-around-ginormous-context-objects/) that plugins then have to unpack.
* __Facade!__ An API should not expose architectural details; similarly, one should not need a deep familiarity with the architecture to add basic plugins. Plugin development and platform enhancement are different developer use cases and should look different.

My concerns about the facade route (or, more generally, on breaking from the "everything's just a plugin and they collaborate" paradigm, and differentiating internal API from our plugin API):

* We no longer benefit (as much) from dogfooding. Currently, we are very regularly exercising the same interfaces and mechanisms for extensibility/modularity that plugins use. We have a lot of insight into how well they work and how to improve them as a consequence.
* Open MCT becomes more monolithic. Encapsulating certain platform features/behaviors as plugins takes a bit more work, but is good for long-term flexibility and maintainability. We don't expect third-party developers to swap out one version of Edit mode for another, but we know that _we_ sometimes do that. Why should that be harder than swapping out plugins?
  * This is also a lesson learned from desktop MCT - late in the project, there were major UX changes around editing/saving that weren't well-encapsulated (IIRC this was deadline-motivated, and the intent was always to refactor later - parallels to NEM and #481.) Project fell into maintenance mode, bugs in new editing/saving model were resolved in-place. By the time we recognized that there were fundamental UX issues in the new design, the scope of a fix for these issues became inestimable, and impractical to resolve incrementally. 
* The specific dependencies of plugins remain unclear; actually, they become _less_ clear than they currently are. You need to read through imperative code to understand a plugin's actual dependencies.
* Sticking a facade on top of the architecture means that any extension points we want to utilize need to get exposed through that facade. This means either hindering ubiquitous extensibility, _or_ creating a big interface (or set of interfaces) that is largely redundant to the architecture we're trying to stick a facade in front of.

Those are my concerns, but I don't know that a facade _necessarily_ implies all those. Facades can be modular and still be facades. There is a spectrum of granularity here: On the coarse end, a plugin depends on Open MCT and interacts with whatever registration facade it exposes; on the fine-grained end, you depend on the specific components which expose the specific extension points you want to use.

In the middle you could have dependencies which specifically present facades for sets of extensions, in logical groupings. Dependencies are made a little more clear, without actually exposing the underlying architecture, and you can easily add more extension points. We can still use the same mechanism for communication across components internally (keep on dogfooding, maintain ubiquitous extensibility), and if we don't want to maintain certain internal-use facades as public API, fine, `@private`.